pkg/secretless/config/v1/config.go

```diff
@@ -0,0 +1,163 @@
+package v1
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+
+	"github.com/go-ozzo/ozzo-validation"
+
+	"gopkg.in/yaml.v2"
+)
+
+// StoredSecret represents not the value of a "secret," but the abstract concept
+// of "a secret stored somewhere".
+//
+// Note that "Name" will be the key that maps to this secret's actual value in
+// the map[string][]byte when the "StoredSecret" itself is looked up by a
+// Resolver.
+type StoredSecret struct {
+	// How client code will refer to the secret's actual value at runtime.
+	// Specifically, the key to the secret's value in the map[string][]byte
+	// returned by a Resolver.
+	Name string
+	Provider string
+	// The identifier within the context of a Provider.  Ie, how a provider
+	// refers to this secret.  Eg, a database primary key.
+	ID string
+}
+
+// Listener listens on a port on socket for inbound connections, which are
+// handed off to Handlers.
+type Listener struct {
+	Address     string
+	CACertFiles []string `yaml:"caCertFiles"`
+	Debug       bool
+	Name        string
+	Protocol    string
+	Socket      string
+}
+
+// Handler processes an inbound message and connects to a specified backend
+// using Credentials which it fetches from a provider.
+type Handler struct {
+	Name         string
+	Type         string
+	ListenerName string `yaml:"listener"`
+	Debug        bool
+	Match        []string         `yaml:"match"`
+	Patterns     []*regexp.Regexp `yaml:"-"`
+	Credentials  []StoredSecret
+}
+
+// Config is the main configuration structure for Secretless.
+// It lists and configures the protocol listeners and handlers.
+type Config struct {
+	Listeners []Listener
+	Handlers  []Handler
+}
+
+// HasCredential indicates whether a Handler has the specified credential.
+func (h Handler) HasCredential(credentialName string) bool {
+	for _, c := range h.Credentials {
+		if c.Name == credentialName {
+			return true
+		}
+	}
+	return false
+}
+
+func NewConfig(buffer []byte) (*Config, error) {
+	config := &Config{}
+	if err := yaml.Unmarshal(buffer, config); err != nil {
+		return nil, err
+	}
+	return config, nil
+}
+
+// Validate verifies the completeness and correctness of the Handler.
+func (h Handler) Validate() (err error) {
+	err = validation.ValidateStruct(&h,
+		validation.Field(&h.Name, validation.Required),
+	)
+
+	return
+}
+
+// SelectHandlers selects the Handlers that are configured
+// to use this Listener.
+func (l Listener) SelectHandlers(handlers []Handler) []Handler {
+	var result []Handler
+	for _, h := range handlers {
```

Comment:
Ditto on old code: `h` -> `handler` or something similar

Reply:
We can table this convo for later but programming languages have moved past C and C++ - there's no need to keep var names short imo 🤷‍♂

Reply:
So as I see it, there are two separate issues here:

1. What is idiomatic Go style?
2. Do we like idiomatic Go Style?  Do we have legitimate reasons to depart from it?

Per the link above (and per Go std lib), the answer to 1. is that short local varnames are favored in Go.   I think we should follow that style for that reason.

The 2nd question is a larger discussion, which would be interesting to have. As you know I am obsessed with naming and clarity.  And fwiw, I fully agree with the Go style here.  I think it's misguided to frame this point as a carryover from a bygone era where crusty sysadmins wrote cryptic one-letter variable names because they were writing only for themselves.  

In fact, there are deep cognitive reasons for preferring short names in local context.

It's actually _less_ readable imo to use verbose variable names _when_ shorter names are clear in context (canonical example: `i` is better than `index` in a short `for` loop).  

Pike and Kernighan make the argument persuasively at the beginning of "The Practice of Programming":

![](https://i.imgur.com/JeUjGkZ.png)